File: shared_resources/src/main/java/com/thewizrd/shared_resources/FirebaseHelper.java
package com.thewizrd.shared_resources;

import android.util.Log;

import com.google.android.gms.tasks.Tasks;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreSettings;
import com.google.firebase.storage.FirebaseStorage;
import com.thewizrd.shared_resources.utils.Logger;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class FirebaseHelper {
    private static boolean sHasSignInFailed;

    public static FirebaseFirestore getFirestoreDB() {
        checkSignIn();
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        db.setFirestoreSettings(new FirebaseFirestoreSettings.Builder()
                .setCacheSizeBytes(FirebaseFirestoreSettings.CACHE_SIZE_UNLIMITED)
                .setPersistenceEnabled(true)
                .setSslEnabled(true)
                .build());
        return db;
    }

    public static FirebaseStorage getFirebaseStorage() {
        checkSignIn();
        return FirebaseStorage.getInstance();
    }

    public static String getAccessToken() {
        checkSignIn();
        final FirebaseAuth auth = FirebaseAuth.getInstance();
        return new AsyncTask<String>().await(new Callable<String>() {
            @Override
            public String call() {
                if (auth.getCurrentUser() != null) {
                    String token = null;
                    try {
                        token = Tasks.await(auth.getCurrentUser().getIdToken(true)).getToken();
                    } catch (Exception e) {
                        Logger.writeLine(Log.DEBUG, "Error getting user token");
                    }

                    return token;
                }
                return null;
            }
        });
    }

    private static void checkSignIn() {
        final FirebaseAuth auth = FirebaseAuth.getInstance();
        if (auth.getCurrentUser() == null && !sHasSignInFailed) {
            new AsyncTask<Void>().await(new Callable<Void>() {
                @Override
                public Void call() {
                    try {
                        Tasks.await(auth.signInAnonymously(), 15, TimeUnit.SECONDS);
                    } catch (ExecutionException | InterruptedException | TimeoutException e) {
                        Logger.writeLine(Log.ERROR, e, "Firebase: failed to sign in");
                        if (e instanceof TimeoutException) {
                            sHasSignInFailed = true;
                        }
                    }
                    return null;
                }
            });
        }
    }
}

File: shared_resources/src/fullgms/java/com/thewizrd/shared_resources/firebase/FirebaseHelper.kt
package com.thewizrd.shared_resources.firebase

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.google.firebase.storage.FirebaseStorage
import com.thewizrd.shared_resources.utils.Logger
import kotlinx.coroutines.*
import kotlinx.coroutines.tasks.await
import java.util.concurrent.TimeUnit

object FirebaseHelper {
    private var sHasSignInFailed = false
    private var sSetupFirebaseDB = false

    suspend fun getFirestoreDB(): FirebaseFirestore {
        checkSignIn()

        val db = FirebaseFirestore.getInstance()
        db.firestoreSettings = FirebaseFirestoreSettings.Builder()
            .setCacheSizeBytes(FirebaseFirestoreSettings.CACHE_SIZE_UNLIMITED)
            .setPersistenceEnabled(true)
            .setSslEnabled(true)
            .build()
        return db
    }

    suspend fun getFirebaseStorage(): FirebaseStorage {
        checkSignIn()

        val stor = FirebaseStorage.getInstance()
        stor.maxDownloadRetryTimeMillis = TimeUnit.HOURS.toMillis(1)
        return stor
    }

    suspend fun getFirebaseDB(): FirebaseDatabase {
        checkSignIn()

        val db = FirebaseDatabase.getInstance()
        if (!sSetupFirebaseDB) {
            db.setPersistenceEnabled(true)
            db.setPersistenceCacheSizeBytes((2 * 1024 * 1024).toLong()) // 2 MB
            sSetupFirebaseDB = true
        }
        return db
    }

    suspend fun getAccessToken(): String? = withContext(Dispatchers.IO) {
        checkSignIn()

        val auth = FirebaseAuth.getInstance()

        var token: String? = null
        try {
            token = auth.currentUser?.getIdToken(true)?.await()?.token
        } catch (e: Exception) {
            Logger.writeLine(Log.DEBUG, "Error getting user token")
        }

        return@withContext token
    }

    private suspend fun checkSignIn() {
        val auth = FirebaseAuth.getInstance()
        if (auth.currentUser == null && !sHasSignInFailed) {
            supervisorScope {
                try {
                    withTimeout(15000 /* 15s */) {
                        auth.signInAnonymously().await()
                    }
                } catch (e: Exception) {
                    if (e is TimeoutCancellationException) {
                        sHasSignInFailed = true
                    }
                    null
                }
            }
        }
    }
}